rust/operator-binary/src/controller.rs

```diff
@@ -0,0 +1,890 @@
+            .await?
+            // TODO: hive now offers all pods fqdn(s) instead of the service
+            //    this should be removed
+            .split('\n')
+            .collect::<Vec<_>>()
+            .into_iter()
+            .next()
+            .map(|s| s.to_string());
+    }
+
+    Ok(hive_connect_string)
+}
+
+/// Defines all required roles and their required configuration.
+///
+/// The roles and their configs are then validated and complemented by the product config.
+///
+/// # Arguments
+/// * `resource`        - The NifiCluster containing the role definitions.
+/// * `version`         - The NifiCluster version.
+/// * `product_config`  - The product config to validate and complement the user config.
+///
+fn validated_product_config(
+    resource: &TrinoCluster,
+    version: &str,
+    product_config: &ProductConfigManager,
+) -> Result<ValidatedRoleConfigByPropertyKind, Error> {
+    let mut roles = HashMap::new();
+
+    let config_files = vec![
+        PropertyNameKind::File(CONFIG_PROPERTIES.to_string()),
+        PropertyNameKind::File(HIVE_PROPERTIES.to_string()),
+        PropertyNameKind::File(NODE_PROPERTIES.to_string()),
+        PropertyNameKind::File(JVM_CONFIG.to_string()),
+        PropertyNameKind::File(LOG_PROPERTIES.to_string()),
+    ];
+
+    roles.insert(
+        TrinoRole::Coordinator.to_string(),
+        (
+            [
+                config_files.clone(),
+                vec![PropertyNameKind::File(
+                    PASSWORD_AUTHENTICATOR_PROPERTIES.to_string(),
+                )],
+            ]
+            .concat(),
+            resource
+                .spec
+                .coordinators
+                .clone()
+                .with_context(|| MissingTrinoRole {
+                    role: TrinoRole::Coordinator.to_string(),
+                })?,
+        ),
+    );
+
+    roles.insert(
+        TrinoRole::Worker.to_string(),
+        (
+            config_files,
+            resource
+                .spec
+                .workers
+                .clone()
+                .with_context(|| MissingTrinoRole {
+                    role: TrinoRole::Worker.to_string(),
+                })?,
+        ),
+    );
+
+    let role_config =
+        transform_all_roles_to_config(resource, roles).context(ProductConfigTransform)?;
+
+    validate_all_roles_and_groups_config(version, &role_config, product_config, false, false)
+        .context(InvalidProductConfig)
+}
+
+async fn cluster_ref_cm_data(
+    client: &Client,
+    cluster_ref: &ClusterRef,
+    product_name: &str,
+) -> Result<String> {
+    let name = &cluster_ref.name;
+    let namespace = &cluster_ref.namespace;
+    Ok(client
+        .get::<ConfigMap>(name, Some(namespace))
+        .await
+        .with_context(|| MissingConfigMap {
+            name: name.clone(),
+            namespace: namespace.clone(),
+        })?
+        .data
+        .and_then(|mut data| data.remove(product_name))
```

Comment:
There's no guarantee that `PRODUCT` will be the useful property for every product, nor even that we will only care about one property at a time.

Reply:
I dont quite get that? Can you elaborate? Its basically a handshake between trino and e.g. hive / opa

Reply:
Yeah, the handshake here is going to have to depend on the product being exposed.

For example, an S3 connection would typically be expressed as a set of properties (API endpoint, region, bucket, access key, secret key) rather than a single URI.